File: frontend/src/types/presentation.ts
export type PresentationVariant = 'legacy' | 'new'
export type PresentationSource = 'global_force_legacy' | 'user_override' | 'tenant_default' | 'product_default'
export type AccentColor = 'cyan' | 'indigo' | 'emerald' | 'rose' | 'amber'
export type FontFamily = 'geist' | 'dm-sans' | 'jakarta' | 'inter'
export type AppearanceFontSize = 'small' | 'default' | 'large'
export type AppearanceDensity = 'compact' | 'default' | 'comfortable' | 'large'
export type NotificationPosition = 'top_right' | 'bottom_right' | 'top_center'
export type AppearanceMode = 'light' | 'dark' | 'high_contrast'

export interface AppearancePreferences {
  accent: AccentColor
  font_family: FontFamily
  font_size: AppearanceFontSize
  density: AppearanceDensity
  notification_position: NotificationPosition
  mode: AppearanceMode
}

export interface PresentationBootstrap {
  schema_version: 1
  resolved_variant: PresentationVariant
  source: PresentationSource
  appearance: AppearancePreferences
  defaults: AppearancePreferences
  revision: number
  legacy_migration_status: 'pending' | 'complete'
  updated_at: string | null
}

export const DEFAULT_APPEARANCE: AppearancePreferences = Object.freeze({
  accent: 'cyan',
  font_family: 'geist',
  font_size: 'default',
  density: 'default',
  notification_position: 'bottom_right',
  mode: 'dark',
})

export function isAppearance(value: unknown): value is AppearancePreferences {
  if (!value || typeof value !== 'object') return false
  const candidate = value as Record<string, unknown>
  return ['cyan', 'indigo', 'emerald', 'rose', 'amber'].includes(String(candidate.accent))
    && ['geist', 'dm-sans', 'jakarta', 'inter'].includes(String(candidate.font_family))
    && ['small', 'default', 'large'].includes(String(candidate.font_size))
    && ['compact', 'default', 'comfortable', 'large'].includes(String(candidate.density))
    && ['top_right', 'bottom_right', 'top_center'].includes(String(candidate.notification_position))
    && ['light', 'dark', 'high_contrast'].includes(String(candidate.mode))
}

export function isPresentationBootstrap(value: unknown): value is PresentationBootstrap {
  if (!value || typeof value !== 'object') return false
  const candidate = value as Record<string, unknown>
  return candidate.schema_version === 1
    && ['legacy', 'new'].includes(String(candidate.resolved_variant))
    && ['global_force_legacy', 'user_override', 'tenant_default', 'product_default'].includes(String(candidate.source))
    && typeof candidate.revision === 'number'
    && candidate.revision >= 0
    && isAppearance(candidate.appearance)
    && isAppearance(candidate.defaults)
    && ['pending', 'complete'].includes(String(candidate.legacy_migration_status))
    && (candidate.updated_at === null || typeof candidate.updated_at === 'string')
}

File: backend/app/schemas/presentation.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


PresentationVariant = Literal["legacy", "new"]
PresentationSource = Literal[
    "global_force_legacy", "user_override", "tenant_default", "product_default"
]
LegacyMigrationStatus = Literal["pending", "complete"]


class AppearanceValues(BaseModel):
    model_config = ConfigDict(extra="forbid")

    accent: Literal["cyan", "indigo", "emerald", "rose", "amber"]
    font_family: Literal["geist", "dm-sans", "jakarta", "inter"]
    font_size: Literal["small", "default", "large"]
    density: Literal["compact", "default", "comfortable", "large"]
    notification_position: Literal["top_right", "bottom_right", "top_center"]
    mode: Literal["light", "dark", "high_contrast"]


class PresentationResponse(BaseModel):
    schema_version: Literal[1] = 1
    resolved_variant: PresentationVariant
    source: PresentationSource
    appearance: AppearanceValues
    defaults: AppearanceValues
    revision: int = Field(ge=0)
    legacy_migration_status: LegacyMigrationStatus
    updated_at: Optional[datetime] = None


class AppearanceUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal[1]
    base_revision: int = Field(ge=0)
    appearance: AppearanceValues
    migration_source: Optional[Literal["legacy_local_v1"]] = None


class AppearanceResetRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal[1]
    base_revision: int = Field(ge=0)


class TenantPresentationRolloutRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal[1]
    presentation: PresentationVariant


class UserPresentationRolloutRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    schema_version: Literal[1]
    presentation_override: Optional[PresentationVariant]


class TenantPresentationRolloutResponse(BaseModel):
    schema_version: Literal[1] = 1
    tenant_id: str
    presentation: PresentationVariant


class UserPresentationRolloutResponse(BaseModel):
    schema_version: Literal[1] = 1
    tenant_id: str
    user_id: str
    presentation_override: Optional[PresentationVariant]
